[BUG] Nextjs build fails when using `useInngestSubscription`

**Describe the bug**
Not sure if this is related to #931 but the Nextjs build fails with the following log. As a workaround I've forced a fallback on next.config file


```typescript
const nextConfig = {
  webpack: (config, {isServer}) => {
    if (!isServer) {
      // Don't resolve 'async_hooks' module on the client
      config.resolve.fallback = {
        ...config.resolve.fallback,
        async_hooks: false,
      }
    }
    return config
  },
  experimental: {
    turbo: {},
  },
}
```

**Code snippets / Logs / Screenshots**
If applicable, add screenshots to help explain your problem.
```shell
> next build

   ▲ Next.js 15.2.3
   - Environments: .env.local
   - Experiments (use with caution):
     · clientTraceMetadata

   Creating an optimized production build ...
<w> [webpack.cache.PackFileCacheStrategy] Serializing big strings (246kiB) impacts deserialization performance (consider using Buffer instead and decode when needed)
Failed to compile.

node:async_hooks
Module build failed: UnhandledSchemeError: Reading from "node:async_hooks" is not handled by plugins (Unhandled scheme).
Webpack supports "data:" and "file:" URIs by default.
You may need an additional plugin to handle "node:" URIs.
    at /Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/webpack/bundle5.js:29:408376
    at Hook.eval [as callAsync] (eval at create (/Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/webpack/bundle5.js:14:9224), <anonymous>:6:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/webpack/bundle5.js:14:6378)
    at Object.processResource (/Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/webpack/bundle5.js:29:408301)
    at processResource (/Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/loader-runner/LoaderRunner.js:1:5308)
    at iteratePitchingLoaders (/Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/loader-runner/LoaderRunner.js:1:4667)
    at runLoaders (/Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/loader-runner/LoaderRunner.js:1:8590)
    at NormalModule._doBuild (/Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/webpack/bundle5.js:29:408163)
    at NormalModule.build (/Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/webpack/bundle5.js:29:410176)
    at /Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/webpack/bundle5.js:29:82494

Import trace for requested module:
node:async_hooks
./node_modules/inngest/components/execution/als.js
./node_modules/inngest/components/execution/v1.js
./node_modules/inngest/components/InngestFunction.js
./node_modules/inngest/components/Inngest.js
./node_modules/inngest/index.js
./node_modules/@inngest/realtime/dist/hooks.js
./app/[locale]/(checkout)/s/[sessionId]/pending/_components/redirect-after-confirmation.tsx
```


**System info (please complete the following information):**

- OS: Macos
- npm package Version: 
  "@inngest/realtime": "^0.1.2",
  "inngest": "^3.34.1",
- Framework  Next.js
- Platform Vercel

Same issue. 


```
> build
> next build

   ▲ Next.js 15.2.4
   - Environments: .env.local
   - Experiments (use with caution):
     · clientTraceMetadata

   Creating an optimized production build ...
<w> [webpack.cache.PackFileCacheStrategy] Serializing big strings (246kiB) impacts deserialization performance (consider using Buffer instead and decode when needed)
Failed to compile.

node:async_hooks
Module build failed: UnhandledSchemeError: Reading from "node:async_hooks" is not handled by plugins (Unhandled scheme).
Webpack supports "data:" and "file:" URIs by default.
You may need an additional plugin to handle "node:" URIs.
    at /Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/webpack/bundle5.js:29:408376
    at Hook.eval [as callAsync] (eval at create (/Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/webpack/bundle5.js:14:9224), <anonymous>:6:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/webpack/bundle5.js:14:6378)
    at Object.processResource (/Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/webpack/bundle5.js:29:408301)
    at processResource (/Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/loader-runner/LoaderRunner.js:1:5308)
    at iteratePitchingLoaders (/Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/loader-runner/LoaderRunner.js:1:4667)
    at runLoaders (/Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/loader-runner/LoaderRunner.js:1:8590)
    at NormalModule._doBuild (/Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/webpack/bundle5.js:29:408163)
    at NormalModule.build (/Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/webpack/bundle5.js:29:410176)
    at /Users/wladpaiva/Developer/i/obvio/node_modules/next/dist/compiled/webpack/bundle5.js:29:82494

Import trace for requested module:
node:async_hooks
./node_modules/@inngest/realtime/node_modules/inngest/components/execution/als.js
./node_modules/@inngest/realtime/node_modules/inngest/components/execution/v1.js
./node_modules/@inngest/realtime/node_modules/inngest/components/InngestFunction.js
./node_modules/@inngest/realtime/node_modules/inngest/components/Inngest.js
./node_modules/@inngest/realtime/node_modules/inngest/index.js
./node_modules/@inngest/realtime/dist/hooks.js
./app/[locale]/(checkout)/s/[sessionId]/pending/_components/redirect-after-confirmation.tsx


> Build failed because of webpack errors
```


Webpack doesnt like the `node:`. If I remove it gives me a different error:


```
> build
> next build

   ▲ Next.js 15.2.4
   - Environments: .env.local
   - Experiments (use with caution):
     · clientTraceMetadata

   Creating an optimized production build ...
<w> [webpack.cache.PackFileCacheStrategy] Serializing big strings (246kiB) impacts deserialization performance (consider using Buffer instead and decode when needed)
Failed to compile.

./node_modules/@inngest/realtime/node_modules/inngest/components/execution/als.js:60:41
Module not found: Can't resolve 'async_hooks'
  58 |     async (resolve) => {
  59 |         try {
> 60 |             const { AsyncLocalStorage } = await Promise.resolve().then(() => __importStar(require("async_hooks")));
     |                                         ^
  61 |             resolve(new AsyncLocalStorage());
  62 |         }
  63 |         catch (err) {

https://nextjs.org/docs/messages/module-not-found

Import trace for requested module:
./node_modules/@inngest/realtime/node_modules/inngest/components/execution/v1.js
./node_modules/@inngest/realtime/node_modules/inngest/components/InngestFunction.js
./node_modules/@inngest/realtime/node_modules/inngest/components/Inngest.js
./node_modules/@inngest/realtime/node_modules/inngest/index.js
./node_modules/@inngest/realtime/dist/hooks.js
./app/[locale]/(checkout)/s/[sessionId]/pending/_components/redirect-after-confirmation.tsx
```

--- a/.changeset/chatty-pandas-know.md
+++ b/.changeset/chatty-pandas-know.md
@@ -1,0 +1,5 @@
+---
+"@inngest/realtime": patch
+---
+
+Public environment variables (such as `NEXT_PUBLIC_*`) are now also accessed when looking for Inngest during subscriptions

--- a/.changeset/early-hotels-doubt.md
+++ b/.changeset/early-hotels-doubt.md
@@ -1,0 +1,5 @@
+---
+"@inngest/realtime": minor
+---
+
+`useInngestSubscription()` no longer accepts an `app`; always use a `token` when subscribing on the client

--- a/.changeset/early-impalas-camp.md
+++ b/.changeset/early-impalas-camp.md
@@ -1,0 +1,5 @@
+---
+"@inngest/realtime": patch
+---
+
+Environment variables are also now captured in Deno and Netlify

--- a/.changeset/rotten-dolphins-try.md
+++ b/.changeset/rotten-dolphins-try.md
@@ -1,0 +1,17 @@
+---
+"@inngest/realtime": minor
+---
+
+`subscribe()` call no longer accepts an `app` as the first parameter
+
+One can be passed alongside other arguments, e.g.
+
+```ts
+subscribe({
+  app,
+  channel: "hello-world",
+  topics: ["messages"],
+});
+```
+
+An app is still required if you are not using a token retrieved from `getSubscriptionToken()`.

--- a/packages/realtime/src/api.ts
+++ b/packages/realtime/src/api.ts
@@ -1,0 +1,80 @@
+import { z } from "zod";
+import { fetchWithAuthFallback, parseAsBoolean } from "./util";
+
+const tokenSchema = z.object({ jwt: z.string() });
+
+export const api = {
+  async getSubscriptionToken({
+    channel,
+    topics,
+    signingKey,
+    signingKeyFallback,
+    apiBaseUrl,
+  }: {
+    channel: string;
+    topics: string[];
+    signingKey: string;
+    signingKeyFallback: string | undefined;
+    apiBaseUrl: string | undefined;
+  }): Promise<string> {
+    let url: URL;
+    const path = "/v1/realtime/token";
+    const inputBaseUrl =
+      apiBaseUrl ||
+      process.env.INNGEST_BASE_URL ||
+      process.env.INNGEST_API_BASE_URL;
+
+    if (inputBaseUrl) {
+      url = new URL(path, inputBaseUrl);
+    } else if (process.env.INNGEST_DEV) {
+      try {
+        const devUrl = new URL(process.env.INNGEST_DEV);
+        url = new URL(path, devUrl);
+      } catch {
+        if (parseAsBoolean(process.env.INNGEST_DEV)) {
+          url = new URL(path, "http://localhost:8288/");
+        } else {
+          url = new URL(path, "https://api.inngest.com/");
+        }
+      }
+    } else {
+      url = new URL(
+        path,
+        process.env.NODE_ENV === "production"
+          ? "https://api.inngest.com/"
+          : "http://localhost:8288/",
+      );
+    }
+
+    const body = topics.map((topic) => ({
+      channel,
+      name: topic,
+      kind: "run",
+    }));
+
+    const res = await fetchWithAuthFallback({
+      authToken: signingKey,
+      authTokenFallback: signingKeyFallback,
+      fetch,
+      url,
+      options: {
+        method: "POST",
+        body: JSON.stringify(body),
+        headers: {
+          "Content-Type": "application/json",
+        },
+      },
+    });
+
+    if (!res.ok) {
+      throw new Error(
+        `Failed to get subscription token: ${res.status} ${
+          res.statusText
+        } - ${await res.text()}`,
+      );
+    }
+
+    const data = await res.json();
+    return tokenSchema.parse(data).jwt;
+  },
+};

--- a/packages/realtime/src/api.ts
+++ b/packages/realtime/src/api.ts
@@ -1,5 +1,5 @@
 import { z } from "zod";
-import { fetchWithAuthFallback, parseAsBoolean } from "./util";
+import { fetchWithAuthFallback, getEnvVar, parseAsBoolean } from "./util";
 
 const tokenSchema = z.object({ jwt: z.string() });
 
@@ -21,17 +21,19 @@ export const api = {
     const path = "/v1/realtime/token";
     const inputBaseUrl =
       apiBaseUrl ||
-      process.env.INNGEST_BASE_URL ||
-      process.env.INNGEST_API_BASE_URL;
+      getEnvVar("INNGEST_BASE_URL") ||
+      getEnvVar("INNGEST_API_BASE_URL");
+
+    const devEnvVar = getEnvVar("INNGEST_DEV");
 
     if (inputBaseUrl) {
       url = new URL(path, inputBaseUrl);
-    } else if (process.env.INNGEST_DEV) {
+    } else if (devEnvVar) {
       try {
-        const devUrl = new URL(process.env.INNGEST_DEV);
+        const devUrl = new URL(devEnvVar);
         url = new URL(path, devUrl);
       } catch {
-        if (parseAsBoolean(process.env.INNGEST_DEV)) {
+        if (parseAsBoolean(devEnvVar)) {
           url = new URL(path, "http://localhost:8288/");
         } else {
           url = new URL(path, "https://api.inngest.com/");
@@ -40,7 +42,7 @@ export const api = {
     } else {
       url = new URL(
         path,
-        process.env.NODE_ENV === "production"
+        getEnvVar("NODE_ENV") === "production"
           ? "https://api.inngest.com/"
           : "http://localhost:8288/",
       );

--- a/packages/realtime/src/hooks.ts
+++ b/packages/realtime/src/hooks.ts
@@ -1,4 +1,3 @@
-import { Inngest } from "inngest";
 import { useEffect, useRef, useState } from "react";
 import { subscribe } from "./subscribe";
 import { type Realtime } from "./types";
@@ -48,18 +47,12 @@ export interface InngestSubscription<TToken extends Realtime.Subscribe.Token> {
 export function useInngestSubscription<
   const TToken extends Realtime.Subscribe.Token | null | undefined,
 >({
-  app = new Inngest({ id: "useInngestSubscription" }),
   token: tokenInput,
   refreshToken,
   key,
   enabled = true,
   bufferInterval = 0,
 }: {
-  /**
-   * TODO
-   */
-  app?: Inngest.Any;
-
   /**
    * TODO
    */
@@ -131,7 +124,7 @@ export function useInngestSubscription<
     const start = async () => {
       try {
         setState(InngestSubscriptionState.Connecting);
-        const stream = await subscribe(app, { ...token });
+        const stream = await subscribe({ ...token });
         if (cancelled) return;
 
         subscriptionRef.current = stream;

--- a/packages/realtime/src/index.test.ts
+++ b/packages/realtime/src/index.test.ts
@@ -342,7 +342,6 @@ describe("subscribe", () => {
       test("can subscribe with just strings", () => {
         const _fn = async () => {
           const stream = await subscribe(
-            app,
             {
               channel: "test",
               topics: ["foo", "bar"],
@@ -389,7 +388,7 @@ describe("subscribe", () => {
     describe("type-only channel import", () => {
       test("errors if channel name is incorrect", () => {
         const _fn = () => {
-          void subscribe(app, {
+          void subscribe({
             // @ts-expect-error Incorrect channel
             channel: typeOnlyChannel<typeof userChannel>("test"),
             topics: ["created", "updated"],
@@ -399,7 +398,7 @@ describe("subscribe", () => {
 
       test("errors if topic names are incorrect with static channel", () => {
         const _fn = () => {
-          void subscribe(app, {
+          void subscribe({
             channel: typeOnlyChannel<typeof staticChannel>("static"),
             // @ts-expect-error Incorrect topic
             topics: ["created", "updated", "test"],
@@ -409,7 +408,7 @@ describe("subscribe", () => {
 
       test("errors if topic names are incorrect with dynamic channel", () => {
         const _fn = () => {
-          void subscribe(app, {
+          void subscribe({
             channel: typeOnlyChannel<typeof userChannel>("user/123"),
             // @ts-expect-error Incorrect topic
             topics: ["created", "updated", "test"],
@@ -420,7 +419,6 @@ describe("subscribe", () => {
       test("can subscribe using types only of a static channel", () => {
         const _fn = async () => {
           const stream = await subscribe(
-            app,
             {
               channel: typeOnlyChannel<typeof staticChannel>("static"),
               topics: ["created", "updated"],
@@ -466,7 +464,6 @@ describe("subscribe", () => {
       test("can subscribe using types only of a dynamic channel", () => {
         const _fn = async () => {
           const stream = await subscribe(
-            app,
             {
               channel: typeOnlyChannel<typeof userChannel>("user/123"),
               topics: ["created", "updated"],
@@ -513,7 +510,7 @@ describe("subscribe", () => {
     describe("runtime channel import", () => {
       test("errors if static definition given", () => {
         const _fn = () => {
-          void subscribe(app, {
+          void subscribe({
             // @ts-expect-error Definition given
             channel: staticChannel,
             topics: ["created", "updated"],
@@ -523,7 +520,7 @@ describe("subscribe", () => {
 
       test("errors if dynamic definition given", () => {
         const _fn = () => {
-          void subscribe(app, {
+          void subscribe({
             // @ts-expect-error Definition given
             channel: userChannel,
             topics: ["created", "updated"],
@@ -533,7 +530,7 @@ describe("subscribe", () => {
 
       test("errors if topic names are incorrect with static channel", () => {
         const _fn = () => {
-          void subscribe(app, {
+          void subscribe({
             channel: staticChannel(),
             // @ts-expect-error Incorrect topic
             topics: ["created", "updated", "test"],
@@ -543,7 +540,7 @@ describe("subscribe", () => {
 
       test("errors if topic names are incorrect with dynamic channel", () => {
         const _fn = () => {
-          void subscribe(app, {
+          void subscribe({
             channel: userChannel("123"),
             // @ts-expect-error Incorrect topic
             topics: ["created", "updated", "test"],
@@ -553,7 +550,7 @@ describe("subscribe", () => {
 
       test("can subscribe with runtime import of a static channel", () => {
         const _fn = async () => {
-          const stream = await subscribe(app, {
+          const stream = await subscribe({
             channel: staticChannel(),
             topics: ["created", "updated"],
           });
@@ -586,7 +583,7 @@ describe("subscribe", () => {
 
       test("can subscribe with runtime import of a dynamic channel", () => {
         const _fn = async () => {
-          const stream = await subscribe(app, {
+          const stream = await subscribe({
             channel: userChannel("123"),
             topics: ["created", "updated"],
           });
@@ -626,7 +623,7 @@ describe("subscribe", () => {
             topics: ["foo", "bar"],
           });
 
-          const stream = await subscribe(app, token, (message) => {
+          const stream = await subscribe(token, (message) => {
             assertType<"test">(message.channel);
             assertType<"foo" | "bar">(message.topic);
 
@@ -670,7 +667,7 @@ describe("subscribe", () => {
             topics: ["created", "updated"],
           });
 
-          const stream = await subscribe(app, token, (message) => {
+          const stream = await subscribe(token, (message) => {
             assertType<"static">(message.channel);
             assertType<"created" | "updated">(message.topic);
 
@@ -714,7 +711,7 @@ describe("subscribe", () => {
             topics: ["created", "updated"],
           });
 
-          const stream = await subscribe(app, token, (message) => {
+          const stream = await subscribe(token, (message) => {
             assertType<"static">(message.channel);
             assertType<"created" | "updated">(message.topic);
 
@@ -758,7 +755,7 @@ describe("subscribe", () => {
             topics: ["created", "updated"],
           });
 
-          const stream = await subscribe(app, token, (message) => {
+          const stream = await subscribe(token, (message) => {
             assertType<`user/${string}`>(message.channel);
             assertType<"created" | "updated">(message.topic);
 
@@ -802,7 +799,7 @@ describe("subscribe", () => {
             topics: ["created", "updated"],
           });
 
-          const stream = await subscribe(app, token, (message) => {
+          const stream = await subscribe(token, (message) => {
             assertType<`user/${string}`>(message.channel);
             assertType<"created" | "updated">(message.topic);
 

--- a/packages/realtime/src/subscribe/TokenSubscription.ts
+++ b/packages/realtime/src/subscribe/TokenSubscription.ts
@@ -1,17 +1,15 @@
 import debug from "debug";
-import { type Inngest } from "inngest";
-import { devServerAvailable, devServerUrl } from "inngest/helpers/devserver";
+import { api } from "../api";
 import { topic } from "../topic";
 import { Realtime } from "../types";
-import { createDeferredPromise } from "../util";
-import { getSubscriptionToken } from "./helpers";
+import { createDeferredPromise, parseAsBoolean } from "../util";
 import { StreamFanout } from "./StreamFanout";
 
 /**
  * TODO
  */
 export class TokenSubscription {
-  #app: Inngest.Any;
+  #apiBaseUrl?: string;
   #channelId: string;
   #debug = debug("inngest:realtime");
   #encoder = new TextEncoder();
@@ -19,6 +17,8 @@ export class TokenSubscription {
   #running = false;
   #topics: Map<string, Realtime.Topic.Definition>;
   #ws: WebSocket | null = null;
+  #signingKey: string | undefined;
+  #signingKeyFallback: string | undefined;
 
   /**
    * This is a map that tracks stream IDs to their corresponding streams and
@@ -33,14 +33,13 @@ export class TokenSubscription {
     /**
      * TODO
      */
-    app: Inngest.Like,
-
-    /**
-     * TODO
-     */
     public token: Realtime.Subscribe.Token,
+    apiBaseUrl: string | undefined,
+    signingKey: string | undefined,
+    signingKeyFallback: string | undefined,
   ) {
-    this.#app = app as Inngest.Any;
+    this.#signingKey = signingKey;
+    this.#signingKeyFallback = signingKeyFallback;
 
     if (typeof token.channel === "string") {
       this.#channelId = token.channel;
@@ -69,26 +68,26 @@ export class TokenSubscription {
     let url: URL;
     const path = "/v1/realtime/connect";
 
-    if (this.#app.apiBaseUrl) {
-      url = new URL(path, this.#app.apiBaseUrl);
-    } else {
-      url = new URL(path, "https://api.inngest.com/");
-
-      if (
-        this.#app["mode"].isDev &&
-        this.#app["mode"].isInferred &&
-        !this.#app.apiBaseUrl
-      ) {
-        const dsUrl = devServerUrl().toString();
-        const devAvailable = await devServerAvailable(
-          dsUrl,
-          this.#app["fetch"],
-        );
-
-        if (devAvailable) {
-          url = new URL(path, dsUrl);
+    if (this.#apiBaseUrl) {
+      url = new URL(path, this.#apiBaseUrl);
+    } else if (process.env.INNGEST_DEV) {
+      try {
+        const devUrl = new URL(process.env.INNGEST_DEV);
+        url = new URL(path, devUrl);
+      } catch {
+        if (parseAsBoolean(process.env.INNGEST_DEV)) {
+          url = new URL(path, "http://localhost:8288/");
+        } else {
+          url = new URL(path, "https://api.inngest.com/");
         }
       }
+    } else {
+      url = new URL(
+        path,
+        process.env.NODE_ENV === "production"
+          ? "https://api.inngest.com/"
+          : "http://localhost:8288/",
+      );
     }
 
     url.protocol = url.protocol === "http:" ? "ws:" : "wss:";
@@ -111,12 +110,31 @@ export class TokenSubscription {
       throw new Error("WebSockets not supported in current environment");
     }
 
-    const key =
-      this.token.key || (await getSubscriptionToken(this.#app, this.token)).key;
+    let key = this.token.key;
     if (!key) {
-      throw new Error(
-        "No subscription token key passed and failed to retrieve one automatically",
+      this.#debug(
+        "No subscription token key passed; attempting to retrieve one automatically...",
       );
+
+      if (!this.#signingKey) {
+        throw new Error(
+          "No subscription token key passed but have no signing key so cannot retrieve one",
+        );
+      }
+
+      key = (
+        await this.lazilyGetSubscriptionToken({
+          ...this.token,
+          signingKey: this.#signingKey,
+          signingKeyFallback: this.#signingKeyFallback,
+        })
+      ).key;
+
+      if (!key) {
+        throw new Error(
+          "No subscription token key passed and failed to retrieve one automatically",
+        );
+      }
     }
 
     const ret = createDeferredPromise<void>();
@@ -382,6 +400,69 @@ export class TokenSubscription {
     }
 
     return ret.promise;
+  }
+
+  /**
+   * TODO
+   */
+  private async lazilyGetSubscriptionToken<
+    const InputChannel extends Realtime.Channel | string,
+    const InputTopics extends (keyof Realtime.Channel.InferTopics<
+      Realtime.Channel.AsChannel<InputChannel>
+    > &
+      string)[],
+    const TToken extends Realtime.Subscribe.Token<
+      Realtime.Channel.AsChannel<InputChannel>,
+      InputTopics
+    >,
+  >(
+    /**
+     * TODO
+     */
+    args: {
+      /**
+       * TODO
+       */
+      channel: Realtime.Subscribe.InferChannelInput<InputChannel>;
+
+      /**
+       * TODO
+       */
+      topics: InputTopics;
+
+      /**
+       * TODO
+       */
+      signingKey: string;
+
+      /**
+       * TODO
+       */
+      signingKeyFallback?: string | undefined;
+    },
+  ): Promise<TToken> {
+    const channelId =
+      typeof args.channel === "string" ? args.channel : args.channel.name;
+
+    if (!channelId) {
+      throw new Error("Channel ID is required to create a subscription token");
+    }
+
+    const key = await api.getSubscriptionToken({
+      channel: channelId,
+      topics: args.topics,
+      signingKey: args.signingKey,
+      signingKeyFallback: args.signingKeyFallback,
+      apiBaseUrl: this.#apiBaseUrl,
+    });
+
+    const token = {
+      channel: channelId,
+      topics: args.topics,
+      key,
+    } as TToken;
+
+    return token;
   }
 
   /**

--- a/packages/realtime/src/subscribe/TokenSubscription.ts
+++ b/packages/realtime/src/subscribe/TokenSubscription.ts
@@ -2,7 +2,11 @@ import debug from "debug";
 import { api } from "../api";
 import { topic } from "../topic";
 import { Realtime } from "../types";
-import { createDeferredPromise, parseAsBoolean } from "../util";
+import {
+  createDeferredPromise,
+  getPublicEnvVar,
+  parseAsBoolean,
+} from "../util";
 import { StreamFanout } from "./StreamFanout";
 
 /**
@@ -38,13 +42,9 @@ export class TokenSubscription {
     signingKey: string | undefined,
     signingKeyFallback: string | undefined,
   ) {
-    this.#apiBaseUrl =
-      apiBaseUrl ||
-      process.env.INNGEST_BASE_URL ||
-      process.env.INNGEST_API_BASE_URL;
-    this.#signingKey = signingKey || process.env.INNGEST_SIGNING_KEY;
-    this.#signingKeyFallback =
-      signingKeyFallback || process.env.INNGEST_SIGNING_KEY_FALLBACK;
+    this.#apiBaseUrl = apiBaseUrl;
+    this.#signingKey = signingKey;
+    this.#signingKeyFallback = signingKeyFallback;
 
     if (typeof token.channel === "string") {
       this.#channelId = token.channel;
@@ -72,15 +72,16 @@ export class TokenSubscription {
   private async getWsUrl(token: string): Promise<URL> {
     let url: URL;
     const path = "/v1/realtime/connect";
+    const devEnvVar = getPublicEnvVar("INNGEST_DEV");
 
     if (this.#apiBaseUrl) {
       url = new URL(path, this.#apiBaseUrl);
-    } else if (process.env.INNGEST_DEV) {
+    } else if (devEnvVar) {
       try {
-        const devUrl = new URL(process.env.INNGEST_DEV);
+        const devUrl = new URL(devEnvVar);
         url = new URL(path, devUrl);
       } catch {
-        if (parseAsBoolean(process.env.INNGEST_DEV)) {
+        if (parseAsBoolean(devEnvVar)) {
           url = new URL(path, "http://localhost:8288/");
         } else {
           url = new URL(path, "https://api.inngest.com/");
@@ -89,7 +90,7 @@ export class TokenSubscription {
     } else {
       url = new URL(
         path,
-        process.env.NODE_ENV === "production"
+        getPublicEnvVar("NODE_ENV") === "production"
           ? "https://api.inngest.com/"
           : "http://localhost:8288/",
       );
@@ -477,7 +478,7 @@ export class TokenSubscription {
     /**
      * TODO
      */
-    reason: string = "Userland closed connection",
+    reason = "Userland closed connection",
   ) {
     if (!this.#running) {
       return;

--- a/packages/realtime/src/subscribe/TokenSubscription.ts
+++ b/packages/realtime/src/subscribe/TokenSubscription.ts
@@ -38,8 +38,13 @@ export class TokenSubscription {
     signingKey: string | undefined,
     signingKeyFallback: string | undefined,
   ) {
-    this.#signingKey = signingKey;
-    this.#signingKeyFallback = signingKeyFallback;
+    this.#apiBaseUrl =
+      apiBaseUrl ||
+      process.env.INNGEST_BASE_URL ||
+      process.env.INNGEST_API_BASE_URL;
+    this.#signingKey = signingKey || process.env.INNGEST_SIGNING_KEY;
+    this.#signingKeyFallback =
+      signingKeyFallback || process.env.INNGEST_SIGNING_KEY_FALLBACK;
 
     if (typeof token.channel === "string") {
       this.#channelId = token.channel;

--- a/packages/realtime/src/subscribe/helpers.ts
+++ b/packages/realtime/src/subscribe/helpers.ts
@@ -1,6 +1,7 @@
 import type { Inngest } from "inngest";
 import type { InngestApi } from "inngest/api/api";
 import type { Realtime } from "../types";
+import { getEnvVar, getPublicEnvVar } from "../util";
 import { TokenSubscription } from "./TokenSubscription";
 
 /**
@@ -46,11 +47,24 @@ export const subscribe = async <
   const app: Inngest.Any | undefined = token.app as Inngest.Any | undefined;
   const api: InngestApi | undefined = app?.["inngestApi"];
 
+  // Allow users to specify public env vars for the target URLs, but do not
+  // allow this for signing keys, as they should never be on a client.
+  const maybeApiBaseUrl =
+    app?.apiBaseUrl ||
+    getPublicEnvVar("INNGEST_BASE_URL") ||
+    getPublicEnvVar("INNGEST_API_BASE_URL");
+
+  const maybeSigningKey =
+    api?.["signingKey"] || getEnvVar("INNGEST_SIGNING_KEY");
+
+  const maybeSigningKeyFallback =
+    api?.["signingKeyFallback"] || getEnvVar("INNGEST_SIGNING_KEY_FALLBACK");
+
   const subscription = new TokenSubscription(
     token as Realtime.Subscribe.Token,
-    app?.apiBaseUrl,
-    api?.["signingKey"],
-    api?.["signingKeyFallback"],
+    maybeApiBaseUrl,
+    maybeSigningKey,
+    maybeSigningKeyFallback,
   );
 
   const retStream = subscription.getJsonStream();

--- a/packages/realtime/src/subscribe/helpers.ts
+++ b/packages/realtime/src/subscribe/helpers.ts
@@ -1,5 +1,5 @@
 import type { Inngest } from "inngest";
-import { InngestApi } from "inngest/api/api";
+import type { InngestApi } from "inngest/api/api";
 import { Realtime } from "../types";
 import { TokenSubscription } from "./TokenSubscription";
 

--- a/packages/realtime/src/subscribe/helpers.ts
+++ b/packages/realtime/src/subscribe/helpers.ts
@@ -1,4 +1,5 @@
-import { Inngest } from "inngest";
+import type { Inngest } from "inngest";
+import { InngestApi } from "inngest/api/api";
 import { Realtime } from "../types";
 import { TokenSubscription } from "./TokenSubscription";
 
@@ -20,12 +21,12 @@ export const subscribe = async <
   /**
    * TODO
    */
-  app: Inngest.Like,
-
-  /**
-   * TODO
-   */
   token: {
+    /**
+     * TODO
+     */
+    app?: Inngest.Like;
+
     /**
      * TODO
      */
@@ -42,9 +43,14 @@ export const subscribe = async <
    */
   callback?: Realtime.Subscribe.Callback<TToken>,
 ): Promise<TOutput> => {
+  const app: Inngest.Any | undefined = token.app as Inngest.Any | undefined;
+  const api: InngestApi | undefined = app?.["inngestApi"];
+
   const subscription = new TokenSubscription(
-    app,
     token as Realtime.Subscribe.Token,
+    app?.apiBaseUrl,
+    api?.["signingKey"],
+    api?.["signingKeyFallback"],
   );
 
   const retStream = subscription.getJsonStream();

--- a/packages/realtime/src/subscribe/helpers.ts
+++ b/packages/realtime/src/subscribe/helpers.ts
@@ -1,6 +1,6 @@
 import type { Inngest } from "inngest";
 import type { InngestApi } from "inngest/api/api";
-import { Realtime } from "../types";
+import type { Realtime } from "../types";
 import { TokenSubscription } from "./TokenSubscription";
 
 /**

--- a/packages/realtime/src/util.ts
+++ b/packages/realtime/src/util.ts
@@ -102,3 +102,95 @@ export const parseAsBoolean = (value: unknown): boolean | undefined => {
 
   return undefined;
 };
+
+const publicEnvVarPrefixes = [
+  "", // Also search for the env var itself
+  "PUBLIC_",
+  "NEXT_PUBLIC_",
+  "REACT_APP_",
+  "NUXT_PUBLIC_",
+  "VUE_APP_",
+];
+
+/**
+ * Given a `key`, get the environment variable under that key.
+ */
+export const getEnvVar = (key: string): string | undefined => {
+  return allProcessEnv()[key];
+};
+
+/**
+ * Given a `key`, get the environment variable under that key or a
+ * public-prefixed version of it, such as `NEXT_PUBLIC_${key}`.
+ */
+export const getPublicEnvVar = (key: string): string | undefined => {
+  const env = allProcessEnv();
+
+  for (const prefix of publicEnvVarPrefixes) {
+    const envVar = env[prefix + key];
+
+    if (envVar !== undefined) {
+      return envVar;
+    }
+  }
+};
+
+export type EnvValue = string | undefined;
+export type Env = Record<string, EnvValue>;
+
+/**
+ * The Deno environment, which is not always available.
+ */
+declare const Deno: {
+  env: { toObject: () => Env };
+};
+
+/**
+ * The Netlify environment, which is not always available.
+ */
+declare const Netlify: {
+  env: { toObject: () => Env };
+};
+
+/**
+ * allProcessEnv returns the current process environment variables, or an empty
+ * object if they cannot be read, making sure we support environments other than
+ * Node such as Deno, too.
+ *
+ * Using this ensures we don't dangerously access `process.env` in environments
+ * where it may not be defined, such as Deno or the browser.
+ */
+export const allProcessEnv = (): Env => {
+  // Node, Bun, or Node-like environments
+  try {
+    if (process.env) {
+      return process.env;
+    }
+  } catch (_err) {
+    // noop
+  }
+
+  // Deno
+  try {
+    const env = Deno.env.toObject();
+
+    if (env) {
+      return env;
+    }
+  } catch (_err) {
+    // noop
+  }
+
+  // Netlify
+  try {
+    const env = Netlify.env.toObject();
+
+    if (env) {
+      return env;
+    }
+  } catch (_err) {
+    // noop
+  }
+
+  return {};
+};

--- a/packages/realtime/src/util.ts
+++ b/packages/realtime/src/util.ts
@@ -42,8 +42,8 @@ export async function fetchWithAuthFallback<TFetch extends typeof fetch>({
   options,
   url,
 }: {
-  authToken?: string;
-  authTokenFallback?: string;
+  authToken: string;
+  authTokenFallback: string | undefined;
   fetch: TFetch;
   options?: Parameters<TFetch>[1];
   url: URL | string;

--- a/packages/realtime/src/util.ts
+++ b/packages/realtime/src/util.ts
@@ -30,3 +30,75 @@ export const createDeferredPromise = <T>(): DeferredPromiseReturn<T> => {
 
   return { promise, resolve: resolve!, reject: reject! };
 };
+
+/**
+ * Send an HTTP request with the given signing key. If the response is a 401 or
+ * 403, then try again with the fallback signing key
+ */
+export async function fetchWithAuthFallback<TFetch extends typeof fetch>({
+  authToken,
+  authTokenFallback,
+  fetch,
+  options,
+  url,
+}: {
+  authToken?: string;
+  authTokenFallback?: string;
+  fetch: TFetch;
+  options?: Parameters<TFetch>[1];
+  url: URL | string;
+}): Promise<Response> {
+  let res = await fetch(url, {
+    ...options,
+    headers: {
+      ...options?.headers,
+      Authorization: `Bearer ${authToken}`,
+    },
+  });
+
+  if ([401, 403].includes(res.status) && authTokenFallback) {
+    res = await fetch(url, {
+      ...options,
+      headers: {
+        ...options?.headers,
+        Authorization: `Bearer ${authTokenFallback}`,
+      },
+    });
+  }
+
+  return res;
+}
+
+/**
+ * Given an unknown value, try to parse it as a `boolean`. Useful for parsing
+ * environment variables that could be a selection of different values such as
+ * `"true"`, `"1"`.
+ *
+ * If the value could not be confidently parsed as a `boolean` or was seen to be
+ * `undefined`, this function returns `undefined`.
+ */
+export const parseAsBoolean = (value: unknown): boolean | undefined => {
+  if (typeof value === "boolean") {
+    return value;
+  }
+
+  if (typeof value === "number") {
+    return Boolean(value);
+  }
+
+  if (typeof value === "string") {
+    const trimmed = value.trim().toLowerCase();
+
+    if (trimmed === "undefined") {
+      return undefined;
+    }
+
+    if (["true", "1"].includes(trimmed)) {
+      return true;
+    }
+
+    return false;
+  }
+
+  return undefined;
+};

